elastic/elasticsearch
# semantic search in ES|QL

tracked in: https://github.com/elastic/elasticsearch/issues/115103

This brings the ability to execute semantic queries using the `match` function on `semantic_text` fields. 




### How to test:

1. Start a trial - the inference API is not available in basic: `POST /_license/start_trial?acknowledge=true`.
2. Create an index using a semantic_text field, here we are using the default ELSER endpoint:
```
PUT testidx
{
  "mappings": {
    "properties": {
      "semantic_text": {
        "type": "semantic_text",
        "inference_id": ".elser-2-elasticsearch"
      }
    }
  }
}
```
3. Ingest some data: this might take a while if ELSER is not yet deployed. To check the status it's useful to have Kibana running, and under `Machine Learning > Trained Models` you can check the status of the ELSER model:
```
POST testidx/_doc
{
  "semantic_text": "live long and prosper"
}

```
4. Query in ES|QL:
```
POST _query?format=txt
{
  "query": """
    FROM testidx
    | WHERE match(semantic_text, "live")
    """
}
```

### Execution

What we do in _search:
```
GET testidx/_search
{
  "query": {
    "bool": {
      "should": {
        "semantic": {
          "field": "semantic_text",
          "query": "live long"
        }
      }
    }
  }
}
```
In _search when we execute a semantic query, we have [2 rewrite phases](https://github.com/elastic/elasticsearch/blob/b42e8f27499b4171c714670dd5c4ab1aa2dd9476/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java#L152-L159):
1. We do an inference call where we send the query string `live long` - the inference results are the embeddings we will use to querying  the `semantic_text` field. Retrieving the embeddings is done on the coordinator, because we do not want to do an inference call on every shard. The inference endpoint might be configured to work with an external, paid service like Azure OpenAI/Amazon Bedrock.
2. On the data node, we do another rewrite depending on the field type- e.g. if the field is configured to work with dense vectors, we rewrite to an exact KNN query.

For ES|QL we kept the same execution flow:

1. On the coordinator, after the logical plan has been analyzed and verified, we look for any `match` function that has a `semantic_text` argument. We collect these semantic query arguments and call the inference service to retrieve the embeddings. We then embed the inference results in the `Match` objects that represent semantic queries(`InferenceUtils`).
2. When we translate the `Match` function to a query builder as part of pushing the filters to Lucene, we do the same type of query rewriting that is done in _search depending on whether the inference results are dense or sparse vectors.I I initially wanted to use `SemanticQueryBuilder` but that resulted in a dependency hell. **NOTE**: this is only temporary, very soon we should have support for `semantic_text` in `MatchQueryBuilder`, so this logic will be simplified.

For licensing:

I did not add any extra licensing checks. After consulting with team, it seems those are not needed.
`SemanticQueryBuilder` also does not do any licensing checks. It relies on the inference APIs to have proper licensing handling.


Checklist:

- [x] CCS is not supported - it's not supported for `semantic` queries in `_search` either - but we need to check we don't fail with a funny error, but with a meaningful message
- [x] it is not possible to query a `semantic_text` field that has different `inference_id`s - validations/tests are needed, this is not supported in `_search` too
